src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,242 @@
+
+        private readonly ImmutableHashSet<string> SystemIOFileMethodMetadataNames = ImmutableHashSet.Create(
+                StringComparer.Ordinal,
+                "WriteAllBytes",
+                "WriteAllLines",
+                "WriteAllText",
+                "Copy",
+                "Move",
+                "AppendAllLines",
+                "AppendAllText",
+                "AppendText",
+                "Delete");
+
+        private readonly ImmutableHashSet<string> SystemReflectionAssemblyMethodMetadataNames = ImmutableHashSet.Create(
+                StringComparer.Ordinal,
+                "GetLoadedModules",
+                "Load",
+                "LoadFile",
+                "LoadFrom",
+                "LoadModule",
+                "LoadWithPartialName",
+                "ReflectionOnlyLoad",
+                "ReflectionOnlyLoadFrom",
+                "UnsafeLoadFrom");
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+                DiagnosticId,
+                s_Title,
+                s_Message,
+                DiagnosticCategory.Security,
+                DiagnosticHelpers.DefaultDiagnosticSeverity,
+                isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+                description: s_Description,
+                helpLinkUri: null,
+                customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            // Security analyzer - analyze and report diagnostics on generated code.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(
+                (CompilationStartAnalysisContext compilationStartAnalysisContext) =>
+                {
+                    var compilation = compilationStartAnalysisContext.Compilation;
+                    var serializableAttributeTypeSymbol = WellKnownTypes.SerializableAttribute(compilation);
+
+                    if (serializableAttributeTypeSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var dangerousMethodSymbolsBuilder = ImmutableArray.CreateBuilder<IMethodSymbol>();
+                    var systemIOFileTypeSymbol = WellKnownTypes.SystemIOFile(compilation);
+
+                    if (systemIOFileTypeSymbol != null)
+                    {
+                        foreach (var targetMethodName in SystemIOFileMethodMetadataNames)
+                        {
+                            dangerousMethodSymbolsBuilder.AddRange(systemIOFileTypeSymbol.GetMembers(targetMethodName).Where(s => s is IMethodSymbol).Cast<IMethodSymbol>());
```

Comment:
Nit: Use `.OfType<IMethodSymbol>()` instead of `.Where(s => s is IMethodSymbol).Cast<IMethodSymbol>()`?